gptel-org.el

```diff
@@ -186,55 +187,64 @@ value of `gptel-org-branching-context', which see."
-                    (gptel--parse-buffer gptel-backend max-entries)))))
-          (display-warning
-             '(gptel org)
-             "Using `gptel-org-branching-context' requires Org version 9.7 or higher, it will be ignored.")
-          (gptel--parse-buffer gptel-backend max-entries))
+        (save-excursion
+          (let* ((org-buf (current-buffer))
+                 (start-bounds (gptel-org--element-lineage-map
+                                   (org-element-at-point) #'gptel-org--element-begin
+                                 '(headline org-data) 'with-self))
+                 (end-bounds
+                  (cl-loop
+                   for (pos . rest) on (cdr start-bounds)
+                   while
+                   (and (>= pos (point-min)) ;respect narrowing
+                        (goto-char pos)
+                        ;; org-element-lineage always returns an extra
+                        ;; (org-data) element at point 1.  If there is also a
+                        ;; heading here, it is either a false positive or we
+                        ;; would be double counting it.  So we reject this node
+                        ;; when also at a heading.
+                        (not (and (eq pos 1) (org-at-heading-p)
+                                  ;; Skip if at the last element of start-bounds,
+                                  ;; since we captured this heading already (#476)
+                                  (null rest))))
+                   do (outline-next-heading)
+                   collect (point) into ends
+                   finally return (cons prompt-end ends))))
+            (with-temp-buffer
+              (setq-local gptel-backend (buffer-local-value 'gptel-backend org-buf)
+                          gptel--system-message
+                          (buffer-local-value 'gptel--system-message org-buf)
+                          gptel-model (buffer-local-value 'gptel-model org-buf)
+                          gptel-mode (buffer-local-value 'gptel-mode org-buf)
+                          gptel-track-response
+                          (buffer-local-value 'gptel-track-response org-buf)
+                          gptel-track-media
+                          (buffer-local-value 'gptel-track-media org-buf))
+              (cl-loop for start in start-bounds
+                       for end in end-bounds
+                       do (insert-buffer-substring org-buf start end)
+                       (goto-char (point-min)))
+              (goto-char (point-max))
+              (gptel--org-unescape-tool-results)
+              (gptel--org-strip-tool-headers)
+              (let ((major-mode 'org-mode))
+                (gptel--parse-buffer gptel-backend max-entries)))))
       ;; Create prompt the usual way
-      (gptel--parse-buffer gptel-backend max-entries))))
+      (let ((source (current-buffer))
+            (beg (point-min))
+            (end (point-max)))
+        (with-temp-buffer
```

Comment:
Since we switched to using a temp buffer for creating prompts on the non-branching path, you'll have to copy over the local variables from the above setq-local block.

While you're at it, perhaps you can make it a little neater?

``` emacs-lisp
(dolist (sym '(gptel-backend gptel--system-message
               gptel-model gptel-mode gptel-track-response
               gptel-track-media))
  (set (make-local-variable sym)
       (buffer-local-value sym org-buf)))
```

Do _not_ try to shortcut this via
``` emacs-lisp
(dolist (sym (buffer-local-variables org-buf))
  (set (make-local-variable sym)
       (buffer-local-value sym org-buf)))
```
as this can have unintended consequences that I'm not sure about, and I'd rather be explicit about the copying right now.

Reply:
I committed something more lame for first pass.  Was about to take a break.

Reply:
https://github.com/karthink/gptel/pull/626/commits/7d8ebdfe494334def8d00871a97ffb7021b3ec51

for squash